File: algorithms/0409.longest_palindrome/longest_palindrome.rs
use std::collections::HashMap;

impl Solution {
    pub fn longest_palindrome(s: String) -> i32 {
        let mut sum = 0;
        let mut alpha_table = HashMap::new();
        for &b in s.as_bytes().iter() {
            let counter = alpha_table.entry(b).or_insert(0);
            *counter += 1;
            sum += 1;
        }
        let odd_number = alpha_table.values().fold(0, |s, &count| {
            if count % 2 == 1 {
                return s + 1;
            }
            s
        });
        if odd_number >= 1 {
            sum = sum - (odd_number - 1);
        }
        sum
    }
}

File: algorithms/0409.longest_palindrome/longest_palindrome.cpp
class Solution {
public:
    int longestPalindrome(const string &s) {
        int odd = 0;
        int sum = 0;
        int alpha_count[256] = {0};
        for (const char &c: s) {
            alpha_count[c] += 1;
        }
        for (const int &n: alpha_count) {
            sum += n;
            if (n & 0x1 == 1) {
                odd = 1;
                sum -= 1;
            }
        }
        return odd + sum;
    }
};